Opioids Overdose Genome Analysis
==============================

## Project Overview
Opioids are now one of the most common causes of accidental death in the US. According to statistics, two out of three drug overdose deaths in 2018 involved an opioid, so opioid abuse can not only affect people physically and mentally but can also deprive their lives (https://docs.google.com/document/d/1JXWb1Bla8iqvyKl3EUxJcGAWBWTvqfO6rRhn1kzrOr4/edit#bookmark=id.p3zzn76i4h3). Opioid addiction has a unique background in that a large reason for why people become addicted is that patients in hospitals are often prescribed opioids to treat pain, however these patients wind up misusing their prescriptions and become addicted.
This is a data science project curated by Cathleen PeÃ±a, Zhaoyi Guo, and Dennis Wu. This github repo contains the codes that are essential to conduct the explicit visualization on the raw data gather from NCBI. 

## Website

The website that introduces the project is under https://genetics.denncc.com/

## Running the Project 

### Pull the Docker Image
To test on the project, in DSMLP, simply pull the image we've generated exclusively for this project by inputting:

    $ launch-180.sh -i dencc/opioids-od:dw -G B04_Genetics
    
### Clone the Repository
In the directory you want to run the project in, run

    $ mkdir temp
    $ cd ./temp 
    $ git clone https://github.com/denncc/opioids-od-genome-analysis

Then you will be able to run the project the project after cloning

### Test the project
To test on the project, simply input

    $ python run.py test

You will be able to see the testing procedure to run

## Project Organization
```
ğŸ“¦opioids-od-genome-analysis
 â”£ ğŸ“‚config
 â”ƒ â”£ ğŸ“œdata_config.json
 â”ƒ â”£ ğŸ“œfeature_config.json
 â”ƒ â”£ ğŸ“œmodel_config.json
 â”ƒ â”£ ğŸ“œsubmission.json
 â”ƒ â”— ğŸ“œtest_config.json
 â”£ ğŸ“‚data
 â”ƒ â”£ ğŸ“‚external
 â”ƒ â”ƒ â”£ ğŸ“‚bam
 â”ƒ â”ƒ â”£ ğŸ“œ.gitkeep
 â”ƒ â”ƒ â”£ ğŸ“œGRCh38_latest_rna.fna
 â”ƒ â”ƒ â”£ ğŸ“œLog.out
 â”ƒ â”ƒ â”£ ğŸ“œSRA_case_table.csv
 â”ƒ â”ƒ â”£ ğŸ“œchrLength.txt
 â”ƒ â”ƒ â”£ ğŸ“œchrName.txt
 â”ƒ â”ƒ â”£ ğŸ“œchrNameLength.txt
 â”ƒ â”ƒ â”£ ğŸ“œchrStart.txt
 â”ƒ â”ƒ â”£ ğŸ“œgencode.v24.annotation.gff3
 â”ƒ â”ƒ â”£ ğŸ“œgencode.v24.annotation.gtf
 â”ƒ â”ƒ â”£ ğŸ“œgencode.v24.annotation_mrna.gff
 â”ƒ â”ƒ â”— ğŸ“œgenomeParameters.txt
 â”ƒ â”£ ğŸ“‚interim
 â”ƒ â”ƒ â”£ ğŸ“œ.gitkeep
 â”ƒ â”ƒ â”£ ğŸ“œcts.tsv
 â”ƒ â”ƒ â”— ğŸ“œdds_res_before_filter.csv
 â”ƒ â”£ ğŸ“‚processed
 â”ƒ â”ƒ â”£ ğŸ“‚duplicates_removed
 â”ƒ â”ƒ â”£ ğŸ“‚htseq
 â”ƒ â”ƒ â”£ ğŸ“‚kallisto
 â”ƒ â”ƒ â”£ ğŸ“‚merged
 â”ƒ â”ƒ â”£ ğŸ“‚sorted
 â”ƒ â”ƒ â”£ ğŸ“‚temp
 â”ƒ â”ƒ â”£ ğŸ“œ.gitkeep
 â”ƒ â”ƒ â”£ ğŸ“œhtseq_cts.csv
 â”ƒ â”ƒ â”£ ğŸ“œhtseq_cts_1.csv
 â”ƒ â”ƒ â”£ ğŸ“œhtseq_cts_gene.csv
 â”ƒ â”ƒ â”£ ğŸ“œhtseq_cts_gene_filtered.csv
 â”ƒ â”ƒ â”£ ğŸ“œkallisto_transcripts.idx
 â”ƒ â”ƒ â”— ğŸ“œtest_gene_counts.csv
 â”ƒ â”£ ğŸ“‚raw
 â”ƒ â”£ ğŸ“‚test
 â”ƒ â”ƒ â”£ ğŸ“‚SRR7949794
 â”ƒ â”ƒ â”ƒ â”£ ğŸ“œabundance.h5
 â”ƒ â”ƒ â”ƒ â”£ ğŸ“œabundance.tsv
 â”ƒ â”ƒ â”ƒ â”£ ğŸ“œpseudoalignments.bam
 â”ƒ â”ƒ â”ƒ â”— ğŸ“œrun_info.json
 â”ƒ â”ƒ â”£ ğŸ“œSRR7949794_1.fastq.gz
 â”ƒ â”ƒ â”— ğŸ“œSRR7949794_2.fastq.gz
 â”ƒ â”— ğŸ“œSRA_case_table.csv
 â”£ ğŸ“‚docs
 â”ƒ â”£ ğŸ“œMakefile
 â”ƒ â”£ ğŸ“œcommands.rst
 â”ƒ â”£ ğŸ“œconf.py
 â”ƒ â”£ ğŸ“œgetting-started.rst
 â”ƒ â”£ ğŸ“œindex.rst
 â”ƒ â”— ğŸ“œmake.bat
 â”£ ğŸ“‚models
 â”ƒ â”— ğŸ“œ.gitkeep
 â”£ ğŸ“‚notebooks
 â”ƒ â”£ ğŸ“œ.gitkeep
 â”ƒ â”£ ğŸ“œEDA_python.ipynb
 â”ƒ â”£ ğŸ“œEDA_r.ipynb
 â”ƒ â”£ ğŸ“œHTSeq.ipynb
 â”ƒ â”£ ğŸ“œSRA_eda.ipynb
 â”ƒ â”— ğŸ“œhtseq_cts.py
 â”£ ğŸ“‚references
 â”ƒ â”— ğŸ“œ.gitkeep
 â”£ ğŸ“‚reports
 â”ƒ â”£ ğŸ“‚figures
 â”ƒ â”ƒ â”£ ğŸ“œ.gitkeep
 â”ƒ â”ƒ â”£ ğŸ“œDist_of_Age.pdf
 â”ƒ â”ƒ â”£ ğŸ“œScatterplot_Matrix_All.pdf
 â”ƒ â”ƒ â”£ ğŸ“œScatterplot_Matrix_Users.pdf
 â”ƒ â”ƒ â”£ ğŸ“œcocaine_use_diff_means.pdf
 â”ƒ â”ƒ â”£ ğŸ“œcocaine_use_means.pdf
 â”ƒ â”ƒ â”£ ğŸ“œdiff_group_means.pdf
 â”ƒ â”ƒ â”£ ğŸ“œdrug_use_pie.pdf
 â”ƒ â”ƒ â”£ ğŸ“œgroup_means.pdf
 â”ƒ â”ƒ â”— ğŸ“œrace_pie.pdf
 â”ƒ â”— ğŸ“œ.gitkeep
 â”£ ğŸ“‚src
 â”ƒ â”£ ğŸ“‚__pycache__
 â”ƒ â”ƒ â”£ ğŸ“œ__init__.cpython-36.pyc
 â”ƒ â”ƒ â”— ğŸ“œ__init__.cpython-37.pyc
 â”ƒ â”£ ğŸ“‚data
 â”ƒ â”ƒ â”£ ğŸ“‚__pycache__
 â”ƒ â”ƒ â”ƒ â”£ ğŸ“œ__init__.cpython-36.pyc
 â”ƒ â”ƒ â”ƒ â”£ ğŸ“œ__init__.cpython-37.pyc
 â”ƒ â”ƒ â”ƒ â”£ ğŸ“œimport_data.cpython-36.pyc
 â”ƒ â”ƒ â”ƒ â”— ğŸ“œimport_data.cpython-37.pyc
 â”ƒ â”ƒ â”£ ğŸ“œ.gitkeep
 â”ƒ â”ƒ â”£ ğŸ“œ__init__.py
 â”ƒ â”ƒ â”£ ğŸ“œ__init__.pyc
 â”ƒ â”ƒ â”£ ğŸ“œimport_data.py
 â”ƒ â”ƒ â”— ğŸ“œmake_dataset.py
 â”ƒ â”£ ğŸ“‚features
 â”ƒ â”ƒ â”£ ğŸ“‚__pycache__
 â”ƒ â”ƒ â”ƒ â”£ ğŸ“œ__init__.cpython-36.pyc
 â”ƒ â”ƒ â”ƒ â”£ ğŸ“œ__init__.cpython-37.pyc
 â”ƒ â”ƒ â”ƒ â”£ ğŸ“œbuild_features.cpython-36.pyc
 â”ƒ â”ƒ â”ƒ â”— ğŸ“œbuild_features.cpython-37.pyc
 â”ƒ â”ƒ â”£ ğŸ“‚r_scripts
 â”ƒ â”ƒ â”ƒ â”— ğŸ“œmain.R
 â”ƒ â”ƒ â”£ ğŸ“œ.gitkeep
 â”ƒ â”ƒ â”£ ğŸ“œ__init__.py
 â”ƒ â”ƒ â”— ğŸ“œbuild_features.py
 â”ƒ â”£ ğŸ“‚models
 â”ƒ â”ƒ â”£ ğŸ“‚__pycache__
 â”ƒ â”ƒ â”ƒ â”£ ğŸ“œ__init__.cpython-37.pyc
 â”ƒ â”ƒ â”ƒ â”£ ğŸ“œbuild_model.cpython-37.pyc
 â”ƒ â”ƒ â”ƒ â”— ğŸ“œhtseq_cts.cpython-37.pyc
 â”ƒ â”ƒ â”£ ğŸ“‚r_scripts
 â”ƒ â”ƒ â”ƒ â”£ ğŸ“œdeseq2.R
 â”ƒ â”ƒ â”ƒ â”£ ğŸ“œvisualization.R
 â”ƒ â”ƒ â”ƒ â”— ğŸ“œwgcna.R
 â”ƒ â”ƒ â”£ ğŸ“‚sh_scripts
 â”ƒ â”ƒ â”ƒ â”— ğŸ“œsamtools.sh
 â”ƒ â”ƒ â”£ ğŸ“œ.Rhistory
 â”ƒ â”ƒ â”£ ğŸ“œ.gitkeep
 â”ƒ â”ƒ â”£ ğŸ“œ__init__.py
 â”ƒ â”ƒ â”£ ğŸ“œbuild_model.py
 â”ƒ â”ƒ â”— ğŸ“œhtseq_cts.py
 â”ƒ â”£ ğŸ“‚visualization
 â”ƒ â”ƒ â”£ ğŸ“œ.gitkeep
 â”ƒ â”ƒ â”£ ğŸ“œ__init__.py
 â”ƒ â”ƒ â”— ğŸ“œvisualize.py
 â”ƒ â”£ ğŸ“œ__init__.py
 â”ƒ â”— ğŸ“œ__init__.pyc
 â”£ ğŸ“œ.gitignore
 â”£ ğŸ“œDockerfile
 â”£ ğŸ“œLICENSE
 â”£ ğŸ“œMakefile
 â”£ ğŸ“œREADME.md
 â”£ ğŸ“œcommand-line-htseq.txt
 â”£ ğŸ“œr-bio.yaml
 â”£ ğŸ“œrequirements.txt
 â”£ ğŸ“œrun.py
 â”£ ğŸ“œsetup.py
 â”£ ğŸ“œtest_environment.py
 â”— ğŸ“œtox.ini
```
